resources/assets/lib/beatmap-discussions/editor.tsx

```diff
@@ -0,0 +1,467 @@
+    }
+
+    return (
+      <span {...props.attributes}>{children}</span>
+    );
+  }
+
+  resetInput = (event?: React.MouseEvent) => {
+    if (event) {
+      event.preventDefault();
+
+      if (!confirm(osu.trans('common.confirmation'))) {
+        return;
+      }
+    }
+
+    Transforms.deselect(this.slateEditor);
+    this.onChange(this.emptyDocTemplate);
+  }
+
+  serialize = (): string => {
+    const review: BeatmapDiscussionReview = [];
+
+    this.state.value.forEach((node: SlateNode) => {
+      switch (node.type) {
+        case 'paragraph':
+          const childOutput: string[] = [];
+          const currentMarks = {
+            bold: false,
+            italic: false,
+          };
+
+          node.children.forEach((child: SlateNode) => {
+            if (child.text !== '') {
+              if (currentMarks.bold !== (child.bold ?? false)) {
+                currentMarks.bold = child.bold;
+                childOutput.push('**');
+              }
+
+              if (currentMarks.italic !== (child.italic ?? false)) {
+                currentMarks.italic = child.italic;
+                childOutput.push('*');
+              }
+            }
+
+            childOutput.push(child.text.replace('*', '\\*'));
+          });
+
+          // ensure closing of open tags
+          if (currentMarks.bold) {
+            childOutput.push('**');
+          }
+          if (currentMarks.italic) {
+            childOutput.push('*');
+          }
+
+          review.push({
+            text: childOutput.join(''),
+            type: 'paragraph',
+          });
+
+          currentMarks.bold = currentMarks.italic = false;
+          break;
+
+        case 'embed':
+          const doc: DocumentIssueEmbed = {
+            beatmap_id: node.beatmapId,
+            discussion_type: node.discussionType,
+            text: node.children[0].text,
+            timestamp: node.timestamp ? BeatmapDiscussionHelper.parseTimestamp(node.timestamp) : null,
+            type: 'embed',
+          };
+
+          if (node.discussionId) {
+            doc.discussion_id = node.discussionId;
+          }
+
+          review.push(doc);
+          break;
+      }
+    });
+
+    return JSON.stringify(review);
+  }
+
+  sortedBeatmaps = () => {
+    if (this.cache.beatmaps) {
+      return this.cache.beatmaps;
+    }
+    this.cache.beatmaps = BeatmapHelper.sort(_.flatten(_.values(this.props.beatmaps)));
+
+    return this.cache.beatmaps;
+  }
+
+  toggleBold = () => {
+    this.toggleMark('bold');
+  }
+
+  toggleEditMode = (enabled: boolean) => {
+    if (!this.props.document || !this.props.discussions || _.isEmpty(this.props.discussions)) {
+      return;
+    }
+
+    if (enabled) {
+      this.setState({
+        value: parseFromMarkdown(this.props.document, this.props.discussions),
+      });
+    }
+  }
+
+  toggleItalic = () => {
+    this.toggleMark('italic');
+  }
+
+  toggleMark = (format: 'bold' | 'italic') => {
```

Comment:
this and `toggleFormat` (why different name?) in `ToolbarButton` behaves slightly differently:

1. create a text with bolded word in the middle (`something **like ** this`)
2. highlight the whole text
3. toggle bold

result:
- bolded text if toggled using mod-b
- unbolded text if toggled using toolbar

Reply:
actually, why are they two different functions?